File: src/test/java/cz/vhromada/catalog/facade/impl/MusicFacadeImplIntegrationTest.java
package cz.vhromada.catalog.facade.impl;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import java.util.Collections;
import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.CatalogTestConfiguration;
import cz.vhromada.catalog.domain.Song;
import cz.vhromada.catalog.entity.Music;
import cz.vhromada.catalog.facade.MusicFacade;
import cz.vhromada.catalog.utils.MusicUtils;
import cz.vhromada.catalog.utils.SongUtils;
import cz.vhromada.common.Time;
import cz.vhromada.common.facade.MovableParentFacade;
import cz.vhromada.common.test.facade.MovableParentFacadeIntegrationTest;
import cz.vhromada.validation.result.Event;
import cz.vhromada.validation.result.Result;
import cz.vhromada.validation.result.Severity;
import cz.vhromada.validation.result.Status;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.test.context.ContextConfiguration;

/**
 * A class represents integration test for class {@link MusicFacadeImpl}.
 *
 * @author Vladimir Hromada
 */
@ContextConfiguration(classes = CatalogTestConfiguration.class)
class MusicFacadeImplIntegrationTest extends MovableParentFacadeIntegrationTest<Music, cz.vhromada.catalog.domain.Music> {

    /**
     * Instance of {@link EntityManager}
     */
    @Autowired
    @Qualifier("containerManagedEntityManager")
    private EntityManager entityManager;

    /**
     * Instance of {@link MusicFacade}
     */
    @Autowired
    private MusicFacade facade;

    /**
     * Test method for {@link MusicFacade#add(Music)} with music with null name.
     */
    @Test
    void add_NullName() {
        final Music music = newData(null);
        music.setName(null);

        final Result<Void> result = facade.add(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_NAME_NULL", "Name mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#add(Music)} with music with empty string as name.
     */
    @Test
    void add_EmptyName() {
        final Music music = newData(null);
        music.setName("");

        final Result<Void> result = facade.add(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_NAME_EMPTY", "Name mustn't be empty string.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#add(Music)} with music with null URL to english Wikipedia about music.
     */
    @Test
    void add_NullWikiEn() {
        final Music music = newData(null);
        music.setWikiEn(null);

        final Result<Void> result = facade.add(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_WIKI_EN_NULL",
                "URL to english Wikipedia page about music mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#add(Music)} with music with null URL to czech Wikipedia about music.
     */
    @Test
    void add_NullWikiCz() {
        final Music music = newData(null);
        music.setWikiCz(null);

        final Result<Void> result = facade.add(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_WIKI_CZ_NULL",
                "URL to czech Wikipedia page about music mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#add(Music)} with music with not positive count of media.
     */
    @Test
    void add_NotPositiveMediaCount() {
        final Music music = newData(null);
        music.setMediaCount(0);

        final Result<Void> result = facade.add(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_MEDIA_COUNT_NOT_POSITIVE", "Count of media must be positive number.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#add(Music)} with music with null note.
     */
    @Test
    void add_NullNote() {
        final Music music = newData(null);
        music.setNote(null);

        final Result<Void> result = facade.add(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_NOTE_NULL", "Note mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#update(Music)} with music with null name.
     */
    @Test
    void update_NullName() {
        final Music music = newData(1);
        music.setName(null);

        final Result<Void> result = facade.update(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_NAME_NULL", "Name mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#update(Music)} with music with empty string as name.
     */
    @Test
    void update_EmptyName() {
        final Music music = newData(1);
        music.setName("");

        final Result<Void> result = facade.update(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_NAME_EMPTY", "Name mustn't be empty string.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#update(Music)} with music with null URL to english Wikipedia about music.
     */
    @Test
    void update_NullWikiEn() {
        final Music music = newData(1);
        music.setWikiEn(null);

        final Result<Void> result = facade.update(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_WIKI_EN_NULL",
                "URL to english Wikipedia page about music mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#update(Music)} with music with null URL to czech Wikipedia about music.
     */
    @Test
    void update_NullWikiCz() {
        final Music music = newData(1);
        music.setWikiCz(null);

        final Result<Void> result = facade.update(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_WIKI_CZ_NULL",
                "URL to czech Wikipedia page about music mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#update(Music)} with music with not positive count of media.
     */
    @Test
    void update_NotPositiveMediaCount() {
        final Music music = newData(1);
        music.setMediaCount(0);

        final Result<Void> result = facade.update(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_MEDIA_COUNT_NOT_POSITIVE", "Count of media must be positive number.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#update(Music)} with music with null note.
     */
    @Test
    void update_NullNote() {
        final Music music = newData(1);
        music.setNote(null);

        final Result<Void> result = facade.update(music);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "MUSIC_NOTE_NULL", "Note mustn't be null.")));
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#getTotalMediaCount()}.
     */
    @Test
    void getTotalMediaCount() {
        final int count = 60;

        final Result<Integer> result = facade.getTotalMediaCount();

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getData()).isEqualTo(count);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#getTotalLength()}.
     */
    @Test
    void getTotalLength() {
        final Time length = new Time(666);

        final Result<Time> result = facade.getTotalLength();

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getData()).isEqualTo(length);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        assertDefaultRepositoryData();
    }

    /**
     * Test method for {@link MusicFacade#getSongsCount()}.
     */
    @Test
    void getSongsCount() {
        final Result<Integer> result = facade.getSongsCount();

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.OK);
            softly.assertThat(result.getData()).isEqualTo(SongUtils.SONGS_COUNT);
            softly.assertThat(result.getEvents()).isEmpty();
        });

        assertDefaultRepositoryData();
    }

    @Override
    protected MovableParentFacade<Music> getFacade() {
        return facade;
    }

    @Override
    protected Integer getDefaultDataCount() {
        return MusicUtils.MUSIC_COUNT;
    }

    @Override
    protected Integer getRepositoryDataCount() {
        return MusicUtils.getMusicCount(entityManager);
    }

    @Override
    protected List<cz.vhromada.catalog.domain.Music> getDataList() {
        return MusicUtils.getMusic();
    }

    @Override
    protected cz.vhromada.catalog.domain.Music getDomainData(final Integer index) {
        return MusicUtils.getMusic(index);
    }

    @Override
    protected Music newData(final Integer id) {
        return MusicUtils.newMusic(id);
    }

    @Override
    protected cz.vhromada.catalog.domain.Music newDomainData(final Integer id) {
        return MusicUtils.newMusicDomain(id);
    }

    @Override
    protected cz.vhromada.catalog.domain.Music getRepositoryData(final Integer id) {
        return MusicUtils.getMusic(entityManager, id);
    }

    @Override
    protected String getName() {
        return "Music";
    }

    @Override
    protected void clearReferencedData() {
    }

    @Override
    protected void assertDataListDeepEquals(final List<Music> expected, final List<cz.vhromada.catalog.domain.Music> actual) {
        MusicUtils.assertMusicListDeepEquals(expected, actual);
    }

    @Override
    protected void assertDataDeepEquals(final Music expected, final cz.vhromada.catalog.domain.Music actual) {
        MusicUtils.assertMusicDeepEquals(expected, actual);
    }

    @Override
    protected void assertDataDomainDeepEquals(final cz.vhromada.catalog.domain.Music expected, final cz.vhromada.catalog.domain.Music actual) {
        MusicUtils.assertMusicDeepEquals(expected, actual);
    }

    @Override
    protected void assertDefaultRepositoryData() {
        super.assertDefaultRepositoryData();

        assertReferences();
    }

    @Override
    protected void assertNewRepositoryData() {
        super.assertNewRepositoryData();

        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(0);
    }

    @Override
    protected void assertAddRepositoryData() {
        super.assertAddRepositoryData();

        assertReferences();
    }

    @Override
    protected void assertUpdateRepositoryData() {
        super.assertUpdateRepositoryData();

        assertReferences();
    }

    @Override
    protected void assertRemoveRepositoryData() {
        super.assertRemoveRepositoryData();

        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(SongUtils.SONGS_COUNT - SongUtils.SONGS_PER_MUSIC_COUNT);
    }

    @Override
    protected void assertDuplicateRepositoryData() {
        super.assertDuplicateRepositoryData();

        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(SongUtils.SONGS_COUNT + SongUtils.SONGS_PER_MUSIC_COUNT);
    }

    @Override
    protected cz.vhromada.catalog.domain.Music getExpectedDuplicatedData() {
        final cz.vhromada.catalog.domain.Music music = super.getExpectedDuplicatedData();
        for (final Song song : music.getSongs()) {
            song.setId(SongUtils.SONGS_COUNT + music.getSongs().indexOf(song) + 1);
        }

        return music;
    }

    /**
     * Asserts references.
     */
    private void assertReferences() {
        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(SongUtils.SONGS_COUNT);
    }

}

File: src/test/kotlin/cz/vhromada/catalog/facade/impl/MusicFacadeImplIntegrationTest.kt
package cz.vhromada.catalog.facade.impl

import cz.vhromada.catalog.CatalogTestConfiguration
import cz.vhromada.catalog.entity.Music
import cz.vhromada.catalog.facade.MusicFacade
import cz.vhromada.catalog.utils.MusicUtils
import cz.vhromada.catalog.utils.SongUtils
import cz.vhromada.common.Time
import cz.vhromada.common.facade.MovableParentFacade
import cz.vhromada.common.test.facade.MovableParentFacadeIntegrationTest
import cz.vhromada.validation.result.Event
import cz.vhromada.validation.result.Severity
import cz.vhromada.validation.result.Status
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.test.context.ContextConfiguration
import javax.persistence.EntityManager

/**
 * A class represents integration test for class [MusicFacadeImpl].
 *
 * @author Vladimir Hromada
 */
@ContextConfiguration(classes = [CatalogTestConfiguration::class])
class MusicFacadeImplIntegrationTest : MovableParentFacadeIntegrationTest<Music, cz.vhromada.catalog.domain.Music>() {

    /**
     * Instance of [EntityManager]
     */
    @Autowired
    @Qualifier("containerManagedEntityManager")
    private lateinit var entityManager: EntityManager

    /**
     * Instance of [MusicFacade]
     */
    @Autowired
    private lateinit var facade: MusicFacade

    /**
     * Test method for [MusicFacade.add] with music with null name.
     */
    @Test
    fun add_NullName() {
        val music = newData(null)
                .copy(name = null)

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_NAME_NULL", "Name mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.add] with music with empty string as name.
     */
    @Test
    fun add_EmptyName() {
        val music = newData(null)
                .copy(name = "")

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_NAME_EMPTY", "Name mustn't be empty string.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.add] with music with null URL to english Wikipedia about music.
     */
    @Test
    fun add_NullWikiEn() {
        val music = newData(null)
                .copy(wikiEn = null)

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_WIKI_EN_NULL",
                    "URL to english Wikipedia page about music mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.add] with music with null URL to czech Wikipedia about music.
     */
    @Test
    fun add_NullWikiCz() {
        val music = newData(null)
                .copy(wikiCz = null)

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_WIKI_CZ_NULL",
                    "URL to czech Wikipedia page about music mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.add] with music with null count of media.
     */
    @Test
    fun add_NullMediaCount() {
        val music = newData(null)
                .copy(mediaCount = null)

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_MEDIA_COUNT_NULL", "Count of media mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.add] with music with not positive count of media.
     */
    @Test
    fun add_NotPositiveMediaCount() {
        val music = newData(null)
                .copy(mediaCount = 0)

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_MEDIA_COUNT_NOT_POSITIVE", "Count of media must be positive number.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.add] with music with null note.
     */
    @Test
    fun add_NullNote() {
        val music = newData(null)
                .copy(note = null)

        val result = facade.add(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_NOTE_NULL", "Note mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with null name.
     */
    @Test
    fun update_NullName() {
        val music = newData(1)
                .copy(name = null)

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_NAME_NULL", "Name mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with empty string as name.
     */
    @Test
    fun update_EmptyName() {
        val music = newData(1)
                .copy(name = "")

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_NAME_EMPTY", "Name mustn't be empty string.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with null URL to english Wikipedia about music.
     */
    @Test
    fun update_NullWikiEn() {
        val music = newData(1)
                .copy(wikiEn = null)

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_WIKI_EN_NULL",
                    "URL to english Wikipedia page about music mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with null URL to czech Wikipedia about music.
     */
    @Test
    fun update_NullWikiCz() {
        val music = newData(1)
                .copy(wikiCz = null)

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_WIKI_CZ_NULL",
                    "URL to czech Wikipedia page about music mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with null count of media.
     */
    @Test
    fun update_NullMediaCount() {
        val music = newData(1)
                .copy(mediaCount = null)

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_MEDIA_COUNT_NULL", "Count of media mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with not positive count of media.
     */
    @Test
    fun update_NotPositiveMediaCount() {
        val music = newData(1)
                .copy(mediaCount = 0)

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_MEDIA_COUNT_NOT_POSITIVE", "Count of media must be positive number.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.update] with music with null note.
     */
    @Test
    fun update_NullNote() {
        val music = newData(1)
                .copy(note = null)

        val result = facade.update(music)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "MUSIC_NOTE_NULL", "Note mustn't be null.")))
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.getTotalMediaCount].
     */
    @Test
    fun getTotalMediaCount() {
        val result = facade.getTotalMediaCount()

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.data).isEqualTo(60)
            it.assertThat(result.events()).isEmpty()
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.getTotalLength].
     */
    @Test
    fun getTotalLength() {
        val result = facade.getTotalLength()

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.data).isEqualTo(Time(666))
            it.assertThat(result.events()).isEmpty()
        }

        assertDefaultRepositoryData()
    }

    /**
     * Test method for [MusicFacade.getSongsCount].
     */
    @Test
    fun getSongsCount() {
        val result = facade.getSongsCount()

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.OK)
            it.assertThat(result.data).isEqualTo(SongUtils.SONGS_COUNT)
            it.assertThat(result.events()).isEmpty()
        }

        assertDefaultRepositoryData()
    }

    override fun getFacade(): MovableParentFacade<Music> {
        return facade
    }

    override fun getDefaultDataCount(): Int {
        return MusicUtils.MUSIC_COUNT
    }

    override fun getRepositoryDataCount(): Int {
        return MusicUtils.getMusicCount(entityManager)
    }

    override fun getDataList(): List<cz.vhromada.catalog.domain.Music> {
        return MusicUtils.getMusic()
    }

    override fun getDomainData(index: Int): cz.vhromada.catalog.domain.Music {
        return MusicUtils.getMusic(index)
    }

    override fun newData(id: Int?): Music {
        return MusicUtils.newMusic(id)
    }

    override fun newDomainData(id: Int): cz.vhromada.catalog.domain.Music {
        return MusicUtils.newMusicDomain(id)
    }

    override fun getRepositoryData(id: Int): cz.vhromada.catalog.domain.Music? {
        return MusicUtils.getMusic(entityManager, id)
    }

    override fun getName(): String {
        return "Music"
    }

    override fun clearReferencedData() {}

    override fun assertDataListDeepEquals(expected: List<Music>, actual: List<cz.vhromada.catalog.domain.Music>) {
        MusicUtils.assertMusicListDeepEquals(expected, actual)
    }

    override fun assertDataDeepEquals(expected: Music, actual: cz.vhromada.catalog.domain.Music) {
        MusicUtils.assertMusicDeepEquals(expected, actual)
    }

    override fun assertDataDomainDeepEquals(expected: cz.vhromada.catalog.domain.Music, actual: cz.vhromada.catalog.domain.Music) {
        MusicUtils.assertMusicDeepEquals(expected, actual)
    }

    override fun assertDefaultRepositoryData() {
        super.assertDefaultRepositoryData()

        assertReferences()
    }

    override fun assertNewRepositoryData() {
        super.assertNewRepositoryData()

        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(0)
    }

    override fun assertAddRepositoryData() {
        super.assertAddRepositoryData()

        assertReferences()
    }

    override fun assertUpdateRepositoryData() {
        super.assertUpdateRepositoryData()

        assertReferences()
    }

    override fun assertRemoveRepositoryData() {
        super.assertRemoveRepositoryData()

        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(SongUtils.SONGS_COUNT - SongUtils.SONGS_PER_MUSIC_COUNT)
    }

    override fun assertDuplicateRepositoryData() {
        super.assertDuplicateRepositoryData()

        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(SongUtils.SONGS_COUNT + SongUtils.SONGS_PER_MUSIC_COUNT)
    }

    override fun getExpectedDuplicatedData(): cz.vhromada.catalog.domain.Music {
        val music = super.getExpectedDuplicatedData()
        for (song in music.songs) {
            song.id = SongUtils.SONGS_COUNT + music.songs.indexOf(song) + 1
        }

        return music
    }

    /**
     * Asserts references.
     */
    private fun assertReferences() {
        assertThat(SongUtils.getSongsCount(entityManager)).isEqualTo(SongUtils.SONGS_COUNT)
    }

}
